Sources/MongoSwift/Operations/ListDatabasesOperation.swift

```diff
@@ -0,0 +1,91 @@
+import mongoc
+
+/// Options to use when listing available databases.
+public struct ListDatabasesOptions: Encodable {
+    /// Optionally indicate whether only names should be returned.
+    /// This is internal and only used for implementation purposes. Users should use `listDatabaseNames` if they want
+    /// only names.
+    internal var nameOnly: Bool?
+
+    /// Convenience constructor for basic construction
+    public init(nameOnly: Bool? = nil) {
+        self.nameOnly = nameOnly
+    }
+}
+
+/// A struct modeling the information returned from the `listDatabases` command about a single database.
+public struct DatabaseSpecification: Codable {
+    /// The name of the database.
+    public let name: String
+
+    /// The amount of disk space consumed by this database.
+    public let sizeOnDisk: Int
+
+    /// Whether or not this database is empty.
+    public let empty: Bool
+}
+
+/// Internal intermediate result of a ListDatabases command.
+internal enum ListDatabasesResults {
+    /// Includes the names and sizes.
+    case specs([DatabaseSpecification])
+
+    /// Only includes the names.
+    case names([String])
+}
+
+/// An operation corresponding to a "listDatabases" command on a collection.
+internal struct ListDatabasesOperation {
+    private let filter: Document?
+    private let session: ClientSession?
+    private let client: MongoClient
+    private let options: ListDatabasesOptions?
+
+    internal init(client: MongoClient,
+                  filter: Document?,
+                  options: ListDatabasesOptions?,
+                  session: ClientSession?) {
+        self.client = client
+        self.filter = filter
+        self.options = options
+        self.session = session
+    }
+
+    internal func execute() throws -> ListDatabasesResults {
+        let cmd: Document = ["listDatabases": 1]
+        var opts = try self.client.encoder.encode(self.options)
+        if let filter = self.filter {
+            if opts == nil {
+                opts = Document()
+            }
+            opts = ["filter": filter] as Document
```

Comment:
will this overwrite `nameOnly` if it was in the `opts`?

Reply:
Yep, good catch. was supposed to be a field assignment, not the whole doc. Due to including filter in the options struct, this all goes away.